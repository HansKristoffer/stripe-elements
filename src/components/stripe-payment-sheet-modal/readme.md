# stripe-payment-sheet-modal

<!-- Auto Generated Below -->

## Properties

| Property                           | Attribute                               | Description                                                                                                                                                                                 | Type                                                      | Default     |
| ---------------------------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------- | ----------- |
| `handleSubmit`                     | --                                      | Form submit event handler                                                                                                                                                                   | `(event: Event, props: FormSubmitEvent) => Promise<void>` | `undefined` |
| `intentClientSecret`               | `intent-client-secret`                  | The client secret from paymentIntent.create response                                                                                                                                        | `string`                                                  | `undefined` |
| `intentType`                       | `intent-type`                           | Default submit handle type. If you want to use `setupIntent`, should update this attribute.                                                                                                 | `"payment" \| "setup"`                                    | `'payment'` |
| `open`                             | `open`                                  | Modal state. If true, the modal will open                                                                                                                                                   | `boolean`                                                 | `false`     |
| `publishableKey`                   | `publishable-key`                       | Your Stripe publishable API key.                                                                                                                                                            | `string`                                                  | `undefined` |
| `shouldUseDefaultFormSubmitAction` | `should-use-default-form-submit-action` | The component will provide a function to call the `stripe.confirmCardPayment`API. If you want to customize the behavior, should set false. And listen the 'formSubmit' event on the element | `boolean`                                                 | `true`      |
| `showCloseButton`                  | `show-close-button`                     | If true, the modal display close button                                                                                                                                                     | `boolean`                                                 | `true`      |
| `showLabel`                        | `show-label`                            | Show the form label                                                                                                                                                                         | `boolean`                                                 | `false`     |
| `stripeDidLoaded`                  | --                                      | Stripe.js class loaded handler                                                                                                                                                              | `(event: StripeLoadedEvent) => Promise<void>`             | `undefined` |

## Events

| Event    | Description | Type               |
| -------- | ----------- | ------------------ |
| `closed` |             | `CustomEvent<any>` |

## Methods

### `destroy() => Promise<void>`

#### Returns

Type: `Promise<void>`

### `getStripePaymentSheetElement() => Promise<HTMLStripePaymentElement>`

#### Returns

Type: `Promise<HTMLStripePaymentElement>`

### `present() => Promise<unknown>`

#### Returns

Type: `Promise<unknown>`

### `setPaymentRequestButton(options: PaymentRequestButtonOption) => Promise<void>`

#### Returns

Type: `Promise<void>`

### `updateProgress(progress: ProgressStatus) => Promise<HTMLStripePaymentElement>`

#### Returns

Type: `Promise<HTMLStripePaymentElement>`

## Dependencies

### Depends on

- [stripe-sheet](../stripe-element-modal)
- [stripe-payment](../stripe-payment-sheet)

### Graph

```mermaid
graph TD;
  stripe-payment-sheet --> stripe-sheet
  stripe-payment-sheet --> stripe-payment
  stripe-sheet --> ion-icon
  stripe-payment --> stripe-payment-request-button
  style stripe-payment-sheet fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
