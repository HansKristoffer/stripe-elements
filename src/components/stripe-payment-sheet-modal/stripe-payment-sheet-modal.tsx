import { Component, Prop, h, Method, Element, Event, EventEmitter } from '@stencil/core';
import { StripeDidLoadedHandler, FormSubmitHandler, ProgressStatus, PaymentRequestButtonOption, IntentType } from '../../interfaces';

@Component({
  tag: 'stripe-payment-sheet',
  styleUrl: 'stripe-payment-sheet-modal.css',
  shadow: false,
})
export class StripePaymentSheet {
  @Element() el: HTMLStripePaymentSheetElement;

  /**
   * Your Stripe publishable API key.
   */
  @Prop() publishableKey: string;

  /**
   * Overwrite the application name that registered
   * For wrapper library (like Capacitor)
   */
  @Prop() applicationName: string;

  /**
   * Show the form label
   */
  @Prop() showLabel = false;

  /**
   * Payment sheet title
   * By default we recommended to use these string
   * - 'Add your payment information' -> PaymentSheet / PaymentFlow(Android)
   * - 'Add a card' -> PaymentFlow(iOS)
   * These strings will translated automatically by this library.
   */
  @Prop() sheetTitle: string;

  /**
   * Submit button label
   * By default we recommended to use these string
   * - 'Pay' -> PaymentSheet
   * - 'Add' -> PaymentFlow(Android)
   * - 'Add card' -> PaymentFlow(iOS)
   * - 'Add a card' -> PaymentFlow(iOS)
   * These strings will translated automatically by this library.
   *
   */
  @Prop() buttonLabel: string;

  /**
   * The client secret from paymentIntent.create response
   *
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *     stripeElement.setAttribute('intent-client-secret', 'dummy')
   *   })
   * ```
   *
   * @example
   * ```
   * <stripe-card-element intent-client-secret="dummy" />
   * ```
   */
  @Prop() intentClientSecret?: string;

  /**
   * The component will provide a function to call the `stripe.confirmCardPayment`API.
   * If you want to customize the behavior, should set false.
   * And listen the 'formSubmit' event on the element
   */
  @Prop() shouldUseDefaultFormSubmitAction = true;

  /**
   * Default submit handle type.
   * If you want to use `setupIntent`, should update this attribute.
   * @example
   * ```
   * <stripe-payment-sheet intent-type="setup" />
   * ```
   */
  @Prop() intentType: IntentType = 'payment';

  /**
   * Form submit event handler
   */
  @Prop({
    mutable: true,
  })
  handleSubmit: FormSubmitHandler;

  /**
   * Stripe.js class loaded handler
   */
  @Prop({
    mutable: true,
  })
  stripeDidLoaded?: StripeDidLoadedHandler;

  /**
   * If true, the modal display close button
   */
  @Prop() showCloseButton = true;

  /**
   * If true, show zip code field
   */
  @Prop() zip = true;

  /**
   * Modal state.
   * If true, the modal will open
   */
  @Prop() open = false;

  /**
   *
   */
  @Event() closed: EventEmitter;

  componentDidLoad() {
    const modal = this.el.querySelector('stripe-sheet');

    modal.addEventListener('close', () => {
      this.closed.emit();
    });
  }

  /**
   * Get the inner component
   */
  @Method()
  public async getStripePaymentSheetElement() {
    return this.el.querySelector('stripe-payment');
  }

  /**
   * open modal
   */
  @Method()
  public async present() {
    this.open = true;

    return new Promise((resolve, reject) => {
      const paymentSheet = this.el.querySelector('stripe-payment');

      paymentSheet.addEventListener('formSubmit', async props => {
        resolve(props);
      });
      this.el.addEventListener('closed', () => reject());
    });
  }

  /**
   * Update Stripe client loading process
   */
  @Method()
  public async updateProgress(progress: ProgressStatus) {
    const paymentSheet = this.el.querySelector('stripe-payment');

    return paymentSheet.updateProgress(progress);
  }

  /**
   * Remove the modal
   */
  @Method()
  public async destroy() {
    const paymentSheet = this.el.querySelector('stripe-payment');

    paymentSheet.remove();
    this.el.remove();
  }

  /**
   *
   * Add payment request button
   */
  @Method()
  public async setPaymentRequestButton(options: PaymentRequestButtonOption) {
    const elements = this.el.getElementsByTagName('stripe-payment');

    if (elements.length < 1) {
      return;
    }

    const paymentSheetElement = elements[0];

    if (!paymentSheetElement) {
      return;
    }

    paymentSheetElement.setAttribute('show-payment-request-button', 'true');
    if (this.applicationName) {
      paymentSheetElement.setAttribute('application-name', this.applicationName);
    }

    paymentSheetElement.setPaymentRequestOption(options);
  }

  render() {
    return (
      <stripe-sheet open={this.open} showCloseButton={this.showCloseButton}>
        <stripe-payment
          showLabel={this.showLabel}
          publishableKey={this.publishableKey}
          intentClientSecret={this.intentClientSecret}
          shouldUseDefaultFormSubmitAction={this.shouldUseDefaultFormSubmitAction}
          handleSubmit={this.handleSubmit}
          stripeDidLoaded={this.stripeDidLoaded}
          intentType={this.intentType}
          zip={this.zip}
          buttonLabel={this.buttonLabel}
          sheetTitle={this.sheetTitle}
          applicationName={this.applicationName}
        ></stripe-payment>
      </stripe-sheet>
    );
  }
}
