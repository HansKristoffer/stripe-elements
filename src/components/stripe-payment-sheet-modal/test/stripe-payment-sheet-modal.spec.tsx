import { newSpecPage } from '@stencil/core/testing';
import { StripePaymentSheetModal } from '../stripe-payment-sheet-modal';

describe('stripe-payment-sheet-modal', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [StripePaymentSheetModal],
      html: `<stripe-payment-sheet-modal></stripe-payment-sheet-modal>`,
    });
    expect(page.root).toEqualHtml(`
      <stripe-payment-sheet-modal>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </stripe-payment-sheet-modal>
    `);
  });
});
