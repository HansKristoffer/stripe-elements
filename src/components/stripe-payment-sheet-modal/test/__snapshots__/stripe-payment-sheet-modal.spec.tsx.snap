// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripe-payment-sheet renders 1`] = `
<stripe-payment-sheet>
  <stripe-sheet showclosebutton="">
    <stripe-payment intenttype="payment" shouldusedefaultformsubmitaction="" zip=""></stripe-payment>
  </stripe-sheet>
</stripe-payment-sheet>
`;
